server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -310,69 +311,88 @@ private static FieldCapabilitiesResponse merge(
-                res.put("unmapped", unmapped.apply(multiTypes));
+                res.put("unmapped", unmapped.apply(resSize > 1));
             }
             responseMap.put(entry.getKey(), Collections.unmodifiableMap(res));
         }
-        return Collections.unmodifiableMap(responseMap);
+    }
+
+    private static void collectResponseMap(
+        Map<String, Map<String, FieldCapabilities.Builder>> responseMapBuilder,
+        Map<String, Map<String, FieldCapabilities>> responseMap
+    ) {
+        for (Map.Entry<String, Map<String, FieldCapabilities.Builder>> entry : responseMapBuilder.entrySet()) {
+            var typeMapBuilder = entry.getValue().entrySet();
+            responseMap.put(entry.getKey(), Collections.unmodifiableMap(capabilities(typeMapBuilder.size(), typeMapBuilder)));
+        }
+    }
+
+    private static Map<String, FieldCapabilities> capabilities(int resSize, Set<Map.Entry<String, FieldCapabilities.Builder>> builders) {
+        boolean multiTypes = resSize > 1;
+        final Map<String, FieldCapabilities> res = Maps.newHashMapWithExpectedSize(resSize);
+        for (Map.Entry<String, FieldCapabilities.Builder> e : builders) {
+            res.put(e.getKey(), e.getValue().build(multiTypes));
+        }
+        return res;
     }
 
     @Nullable
-    private static Function<Boolean, FieldCapabilities> getUnmappedFields(Set<String> indices, String field, Set<String> mappedIndices) {
-        if (mappedIndices.size() != indices.size()) {
-            return mt -> new FieldCapabilities(
-                field,
-                "unmapped",
-                false,
-                false,
-                false,
-                false,
-                null,
-                mt ? Sets.difference(indices, mappedIndices).toArray(Strings.EMPTY_ARRAY) : null,
-                null,
-                null,
-                null,
-                null,
-                Map.of()
-            );
+    private static Function<Boolean, FieldCapabilities> getUnmappedFields(String[] indices, String field, Set<String> mappedIndices) {
+        if (mappedIndices.size() != indices.length) {
+            return mt -> {
+                final String[] diff;
+                if (mt) {
+                    diff = new String[indices.length - mappedIndices.size()];
+                    Iterator<String> indicesIter = Iterators.forArray(indices);
+                    for (int i = 0; i < diff.length; i++) {
+                        diff[i] = nextIndex(indicesIter, mappedIndices);
+                    }
+                } else {
+                    diff = null;
+                }
+                return new FieldCapabilities(field, "unmapped", false, false, false, false, null, diff, null, null, null, null, Map.of());
+            };
         }
         return null;
     }
 
+    private static String nextIndex(Iterator<String> iter, Set<String> filtered) {
+        while (true) {
+            String index = iter.next();
```

Comment:
Maybe assert `it.hasNext()`?

Reply:
Hmm in theory yes, but now I benchmarked this already and it inlined so so nicely :D I'd rather not increase the code size here I think.